### ISSUE TITLE
Processing UI refactoring

### ISSUE THREAD

**Issue by arnaud-morvan** [qgis/QGIS-Enhancement-Proposals#65]
# QGIS Enhancement: Processing UI refactoring

**Date** 2016/06/17

**Author** Arnaud Morvan (@arnaud-morvan)

**Contact** arnaud dot morvan at camptocamp dot com

**maintainer** @arnaud-morvan

**Version** QGIS 3.0
# Summary
- Add a metadata attribute in parameters to allow using custom widgets in dialogs.
- Move widget initialization code from parameters panels to widget wrappers.

To discuss :
- Emit signals from parameters when value change.
## Current situation

Actually in processing all parameter's widgets are hardcoded in
`ParametersPanel`, depending on parameter types. This is not flexible to
create custom parameters and related panels.

On top of this, we have a lot of code duplicated between:
- Standard algorithm dialog/panel
- Batch dialog/panel
- Modeler dialog/panel
## Proposed Solution

A more modular vision could be to create a widget wrapper class for each parameter type.
ParametersPanel would only have to iterate over parameters and use appropriate wrapper to create the widget, get and set the value.

Using this, it could be easy to create new parameter types and related widget wrappers
without creating a custom dialog and panel class for each case.

Proposed design:
- Parameters will have a new `metadata` attribute.
- `metadata` attribute will contains a `widget_wrapper` property to choice the widget class to display in dialogs and optionally some extra options.
- `widget_wrapper` property will contain a string with the widget wrapper class path for dynamic import in dialogs (as we do not want to import ui classes in core ones).
- Each existing parameter class will have a default `widget_wrapper` value, to keep compatibility with existing algorithms.
- Parameter's constructors would allow to override the default `metadata` values, including the `widget_wrapper` one, in `GeoAlgorithm.defineCharacteristics`.

For a working example see : https://github.com/qgis/QGIS/pull/3449

Benefits :
- We could use custom widgets for specific purpose without creating new parameter classes or altering the common ui classes (ParametersPanel, BatchDialog, ModelerDialog).
- Each processing plugin could provide it's own parameters types and related widgets.
### Example(s)

For example, in postgis related algorithms, it could be possible to create custom widgets, capable of database introspection using existing GeoDB class to propose drop down lists of schemas and tables or primary keys and geometry column detection, without altering existing parameter types. 

Here is an example of ParameterString class definition:

```
class ParameterBoolean(Parameter):

    default_metadata = {
        'widget_wrapper': 'processing.gui.widget_wrappers.BooleanWidgetWrapper'
    }

    def __init__(self, name='', description='', default=None, multiline=False,
                optional=False, evaluateExpressions=False, metadata={}):
```

Here is the widget wrapper class prototype:

```
class WidgetWrapper():

    def __init__(self, param, dialog=DIALOG_STANDARD, extra_values=[])

    def createWidget(self)

    def setValue(self, value)

    def value(self)
```

And a custom parameter instantiation, for a postgis table entry :

```
self.addParameter(ParameterString(self.TABLE,
                                  self.tr('table name'),
                                  'public',
                                  optional=True,
                                  metadata={
                                      'widget_wrapper': {
                                          class: 'processing.gui.PostgisTableWidgetWrapper',
                                          schema_parameter: self.SCHEMA
                                      }
                                  }))
```
### Affected Files

QGIS/python/plugins/processing/core/parameters.py
QGIS/python/plugins/processing/gui/*
## Performance Implications

This should not have significative implication on performance.
## Further Considerations/Improvements

Adding a parameter's `valueChanged` signal is an option. Actually, processing dialogs update parameters value on dialog acceptation. To get a parameter signal when value change, we need that each widget update it's parameter's value in real time.

Here is a view of what we could get with this new strategy :
### Dependant fields functionnality

Currently we have the possibility to give, in ParameterField, the parent layer parameter. Actually, the fields list update is hard coded in `ParametersPanel`.

In field widget, we could get the parent parameter related widget and bind the parent **widget** signal to the fields list update slot.

If we add the `valueChanged` signal on parameters, we could bind the parent **parameter** signal to the fields list update slot.
### ParameterVector QgsVectorLayer cache

In ParameterVector class we could think to a `getLayer` method, responsible to create a QgsVectorLayer instance when needed (when value is a source uri, not a layer instance). In this case, even with a shapefile given by file path, we could get the fields list.
### List of output fields

If we add the `valueChanged` signal on parameters, we can also think to list of output fields getter in GeoAlgorithm class.
## Backwards Compatibility

Adding the `metadata` attribute with `widget` property should not break any compatibility.

Porting widget's initialization code from `ParametersPanel` to parameter's widgets could break custom dialogs depending on existing widgets.

Nothing here should break compatibility with existing algorithms, models or scripts.
## Issue Tracking ID(s)

_(optional)_
## Votes

(required)

**Comment by volaya** [comment]
Nice QEP. This will clean up the code and have nice advantages.

However, there are some things that are being simplified in the document. The main one is that it seems your are assuming that for each parameter there is a widget and the code is duplicated in parameters panel, batch interface, modeler, etc. IT's not like that, those widgets are actually different. So, for the new desing to be backwards compatible, each parameters should provide 3 widgets (standard, batch and modeler)

Other than that, it looks good.

Since this might be a large task..may I ask how much time you would be able to spend on this? I will be available for working on it, and I guess @alexbruy also will be.

Thanks for this nice proposal!

**Comment by mbernasocchi** [comment]
really nice idea, adding a new widget is indeed a bit cumbersome at the moment

**Comment by arnaud-morvan** [comment]
@volaya : instead of creating classes derived from QWidget, we could create widget wrappers with a ``widget`` property and methods to create the widget, access the widget value, etc. Like this we could have only one wrapper for each parameter type.

**Comment by arnaud-morvan** [comment]
For information : as @m-kuhn said, QGIS editor widgets already use widget wrappers with a registry. IMHO we don't need a registry here as we don't need to persist the widgets to use in project file like for editor widgets.

**Comment by arnaud-morvan** [comment]
I've update my QEP with the widget wrapper concept, and I've made a fully working exemple with BooleanWidgetWrapper here :
https://github.com/arnaud-morvan/QGIS/tree/processing_widget_wrappers

**Comment by rkanavath** [comment]
if a parameter changed how do we update the ui (parameter widgets and its values) in a generic way for all toolboxes in processing?.

**Comment by rkanavath** [comment]
a change in one parameter can add/remove widgets inside the processing ui. so basically a single change forces to recreate the widgets inside ui. btw, this recreating of all widgets is not for all parameter only those are explicitly marked in the xml descriptors

**Comment by arnaud-morvan** [comment]
If I correctly understand you request, you'd like some widgets to hide or show depending on some other parameter value. For now you have to make a dedicated dialog for your algorithm.

This seems to be out of the scope of this QEP, but to add this feature in a generic way, it should be necessary to : 
* update the parameter value from a widget value changed signal (it not the case for now, only on OK button click)
* emit a signal from the parameter when the value change
* catch the signal in algorithm and set the other parameters visibility
* emit a signal from the parameter when the visibility change
* catch this signal in the dialogs (standard, batch ?, modeler) and set the visibility on the widget

This could be possible for pure python algorithms but I don't see how to specify this dynamic behavior in a xml description.

Maybe it should be possible to specify dedicated dialogs in xml descriptors, extending from base processing parameter dialogs.

Maybe open a thread on the list.

**Comment by arnaud-morvan** [comment]
I've worked on the ModelerParametersDialog using a dedicated wrapper. It simplify the source code as the common widgets are exactly the same in all dialogs. This has an effect on the user interface as now we have a dedicated button with menu for model specific value (inputs and internal algorithm outputs). IMHO this is easier to understand from a user point of view.

I've opened a pull request if someone wants to test : https://github.com/qgis/QGIS/pull/3449

![modeler_wrapper](https://cloud.githubusercontent.com/assets/6095494/18216536/a2cfaf64-7157-11e6-976c-d6b518cbd94a.png)
